File: java/Game/Android/app/src/main/java/com/example/chessclassic/game/board/piece/Piece.java
package com.example.chessclassic.game.board.piece;

import com.example.chessclassic.game.board.Board;
import com.example.chessclassic.game.board.move.Move;
import com.example.chessclassic.game.player.Colour;

import java.util.List;

import static java.lang.Character.toUpperCase;


public abstract class Piece {
    public int pos;
    public boolean firstMove;
    public Colour colour;
    public PieceType pieceType;

    public Piece(int pos, Colour colour, boolean firstMove) {
        this.pos = pos;
        this.colour = colour;
        this.firstMove = firstMove;
    }

    public static PieceType getPieceTypeByNotation(char notation){
        for (PieceType pieceType : PieceType.values())
            if (pieceType.getCode() == toUpperCase(notation))
                return pieceType;
        return null;
    }
    public static Piece getPieceByPieceType(PieceType pieceType, int position, Colour colour){
        switch(pieceType){
            case PAWN: return new Pawn(position, colour);
            case BISHOP: return new Bishop(position, colour);
            case KNIGHT: return new Knight(position, colour);
            case ROOK: return new Rook(position, colour);
            case QUEEN: return new Queen(position, colour);
            case KING: return new King(position, colour);
            default: return null;
        }
    }

    public abstract Piece copy();
    public abstract List<Move> getPossibleMoves(Board board);

}

File: cpp/game/board/piece/piece.cpp
#include "piece.hpp"

Piece::Piece(bool firstMove, PieceColor color) {
    this->color = color;
    this->firstMove = firstMove;
}
PieceType Piece::getPieceTypeByNotation(char notation) {
    switch (tolower(notation)) {
        case 'p': return Pawn;
        case 'n': return Knight;
        case 'b': return Bishop;
        case 'r': return Rook;
        case 'q': return Queen;
        case 'k': return King;
        default:  return Empty;
    }
}
Piece Piece::getPieceByPieceType(PieceType pieceType, PieceColor color) {
    switch(pieceType){
        case Pawn:   return new Pawn(color);
        case Bishop: return new Bishop(color);
        case Knight: return new Knight(color);
        case Rook:   return new Rook(color);
        case Queen:  return new Queen(color);
        case King:   return new King(color);
        default:     return NULL;
    }
}